interfaces/repo.go

```diff
@@ -516,3 +516,110 @@ func (r *Repository) securitySnippetsForSnap(snapName string, securitySystem Sec
 	}
 	return snippets, nil
 }
+
+// AddSnap adds plugs and slots declared by the given snap to the repository.
+//
+// This function can be used to implement snap install or, when used along with
+// RemoveSnap, snap upgrade.
+//
+// Plugs/slots that are already present in the repository are not changed. Only
+// new plugs/slots are added.
+//
+// Each added plug/slot is validated according to the corresponding interface.
+// Unknown interfaces and plugs/slots that don't validate are not added.
+// Information about those failures are returned to the caller.
+func (r *Repository) AddSnap(snapInfo *snap.Info) []error {
```

Comment:
What will the call site do with such a list of errors?  How does it pop the error up to its caller?

The better way of doing this is by defining a local aggregated error type that renders the message in a pleasant way. Grep to code for ensureError for an example of this pattern.